Refactor Release Notes Script: Improved Workflow

kmelve opened the pull request:
# Summary
This PR refactors the release notes template generator script to improve the workflow, ensure Node.js compatibility, and enhance the documentation. The script now uses a more structured approach for generating release notes and changelog entries.


# Key Changes

## Code Structure and Dependencies
- Converted from ES module imports to CommonJS requires for better compatibility
- Added proper TypeScript type annotations
- Structured the code with clear function separation and error handling

## Functionality Enhancements
- Added support for creating Sanity documents with changelog entries
- Implemented automatic categorization of commits into features and bugfixes
- Added dry run mode for template generation without document creation
- Added debug mode for inspecting document structure

## Template Generation
- Changed from a static template to a dynamically generated one based on commit history
- Added support for custom title, product tag, and version specification
- Improved the GitHub release template to be more concise with links to detailed changelog

## Authentication and Configuration
- Hardcoded project ID and dataset for consistency
- Added support for retrieving auth token from Sanity CLI config
- Added option to provide a direct token via command line

## Documentation
- Created comprehensive README with usage examples and options
- Added documentation on authentication, workflow, and URL formats
- Included examples of command usage with different flags

## Testing
- Verified script execution with Node.js v23
- Tested in dry run mode to ensure proper template generation
- Validated document creation functionality
- Confirmed URL generation for both public-facing and internal Studio links

## Benefits
- **Improved Developer Experience**: Clearer, more structured release notes generation
- **Better Compatibility**: Works with latest Node.js versions without additional configuration
- **Enhanced Documentation**: Comprehensive guide for using the script
- **Streamlined Workflow**: Automated creation of changelog entries in Sanity Studio

## Technical Details
The script now follows a more modular approach with separate functions for different aspects of the release notes generation process. It uses CommonJS requires for better compatibility with different Node.js versions and provides a more robust error handling mechanism.

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,447 @@
+    try {
+      changelogId = await createSanityChangelogDocuments(sanityConfig, releaseContext, flags.debug)
+      // Internal Studio URL for editing (shown in console)
+      const studioUrl = `https://admin.sanity.io/structure/docs;changelog;apiChange;${changelogId}?perspective=rOVi7MsdW`
```
This won't work after the `rOVi7MsdW` release is published, should it make a new release?

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,447 @@
+      description: 'Title for the release',
+      default: '',
+    })
+    .option('product', {
```
Are we currently releasing other products from this repo?

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,447 @@
+To upgrade to this version, run:
+
+\`\`\`bash
+pnpm add sanity@latest
```
we shouldn't assume everyone uses `pnpm`

kmelve commented:
@bjoerge Thanks for the feedback, and shame on me for not properly reviewing AI vibe code. But I have done a bit of refactoring. 

* The changelog entry is supposed to be _edited_ from the raw commit messages that's inserted.
* From now on, we can copy-paste the github release note from the studio.
* I didn't figure out how to automate the semver version, so for now it has to be inserted manually.
* I have noticed that we have "Notes for release" in the PRs, but these are not surfaced by the old or this new releaseNote script. Maybe there is a way to get those into the mix? 

## Changes

1. Added GitHub release notes template as a field in the `apiChange` document:
   - Added `githubReleaseNote` field that includes:
     - Brief introduction
     - Link to full changelog
     - Installation instructions for multiple package managers (npm, pnpm, yarn, and bun)
     - Link to upgrade guide
     - Full commit log

2. Improved TypeScript type safety:
   - Added `PortableTextSpan` and `PortableTextBlock` interfaces for proper typing of Portable Text structures
   - Added `as const` type assertions for `_type` fields to ensure literal type checking
   - Explicitly typed the return value of `generateLongDescription`

3. Code cleanup:
   - Removed redundant console logging of GitHub release template
   - Simplified dry run output handling

## Remaining Issues

- There is a linter error indicating that `SANITY_WEB_AUTH_TOKEN` needs to be added to the root `turbo.json` file's `globalEnv` array, which it is. This should be addressed in a separate PR.

## Testing

The script can be tested using:

```bash
# Dry run to verify document structure
node --no-warnings --experimental-strip-types scripts/printReleaseNotesTemplate.ts --dryRun
# Create actual documents
node --no-warnings --experimental-strip-types scripts/printReleaseNotesTemplate.ts
```

kmelve commented:
> @kmelve could you do a rebase against next and update the PR title so it matches semantic release? 🙏 ref: https://github.com/sanity-io/sanity/actions/runs/13712597273/job/38351775490?pr=8826
> 
> I think the other CI failures are flake and will go away after a rebase

done!

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,553 @@
+    } else {
+      // Fallback to basic keyword matching only if not a conventional commit
+      const lowerMessage = message.toLowerCase()
+      if (lowerMessage.includes('feat') || lowerMessage.includes('feature')) {
```
this will falsely mark commits with "feather" and "defeat" in the message as features

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,553 @@
+      const lowerMessage = message.toLowerCase()
+      if (lowerMessage.includes('feat') || lowerMessage.includes('feature')) {
+        features.push(commit)
+      } else if (lowerMessage.includes('fix') || lowerMessage.includes('bug')) {
```
similarily, this will match any string with "fix" in it, including "prefix", "suffix", etc

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,553 @@
+      }
+      // Other types (chore, docs, etc.) are ignored
+    } else {
+      // Fallback to basic keyword matching only if not a conventional commit
```
Given that we have CI step that enforces conventional commits, I think we should throw if we encounter a commit that doesnt match the spec, as it indicates something is wrong with our setup, which is something we want to know about and fix.

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,553 @@
+            _key: `${blockKey}0`,
+            _type: 'span' as const,
+            marks: [],
+            text: message.replace(/^feat(\([^)]*\))?:\s*/, ''),
```
It would be preferable here if context included the parsed semantic commits instead of raw messages so we don't need to maintain two separate regexes for this

bjoerge commented on scripts/printReleaseNotesTemplate.ts:
```diff
@@ -1,55 +1,553 @@
+            _key: `${blockKey}0`,
+            _type: 'span' as const,
+            marks: [],
+            text: message.replace(/^fix(\([^)]*\))?:\s*/, ''),
```
same as above – would be a lot better if we had parsed commits here instead

bjoerge reviewed (changes requested):
This looks a lot better now, thanks @kmelve!

I still think it's worth outsourcing the hand-written semantic commit parsing to  the `conventional-commits-parser` package, esp. since we already have existing tooling that enforces that all PRs follows it.

bjoerge commented on scripts/README.md:
```diff
@@ -0,0 +1,125 @@
+
+# Specify custom from/to tags and title
+node --no-warnings --experimental-strip-types scripts/printReleaseNotesTemplate.ts \
+  --from v3.15.0 \
```
are --from and --to still supported?

bjoerge commented on scripts/README.md:
```diff
@@ -0,0 +1,125 @@
+
+- Node.js 22+ (for direct TypeScript execution)
+- Sanity authentication token
+- Git repository with commit history
```
Are there any value in adding this? Isn't it implied...?

bjoerge commented on scripts/README.md:
```diff
@@ -0,0 +1,125 @@
+
+### Options
+
+- `--from`: Previous release tag (defaults to the latest tag)
```
--from and --to are removed now, right?
